### PR TITLE
Change order of text processing in log manipulation

### LINKED ISSUE #98: Unicode decode error in logs
Special characters such as "███" (i.e. loading bars and emoji) are being partially removed by the ANSI cleaner, leading to dodgy unicode character filtering into logs which can't be decoded. The solution: a) re-order the text processing so that ANSI is removed on decoded text and b) add tests in to cover "difficult" characters

### PR DESCRIPTION
**Pull request by jaklinger**
Closes #98

> Special characters such as "███" (i.e. loading bars and emoji) are being partially removed by the ANSI cleaner, leading to dodgy unicode character filtering into logs which can't be decoded. The solution: a) re-order the text processing so that ANSI is removed on decoded text and b) add tests in to cover "difficult" characters

If you're feeling lazy, just check the logs [here](https://s3.console.aws.amazon.com/s3/object/open-jobs-lake?region=eu-west-1&prefix=metaflow/metaflow-logs/1628085962/logs.txt) to see that the special character ██ is processed correctly.

As with #92, to test do the following:

## Basic local integration testing

In terms of running the integration test (which is only manual, I don't have a procedure for doing this automatically since it depends on AWS credentials and requires physical infrastructure at present, though could set up mock S3 infrastructure), do the following from `daps_utils/tests/`:

```
In [1]: from test_docker import integration_decode_logs
   ...: integration_decode_logs()

>>> 'See full logs at https://s3.console.aws.amazon.com/s3/object/open-jobs-lake?prefix=metaflow/metaflow-logs/1627391239/logs.txt'
>>> 'first'
>>> 'second'
>>> '... truncated ...'
>>> 'sixth'
>>> 'seventh'
>>> ''
```

And correspondingly if you go to https://s3.console.aws.amazon.com/s3/object/open-jobs-lake?prefix=metaflow/metaflow-logs/1627391239/logs.txt you should see the untruncated logs

## Less basic local integration testing

If you run the local integration tests found in `daps_utils/flows/tests/integration` (i.e. just run `pytest` in that folder) then after a few mins you will find the full metaflowtask logs on S3. For example see [this log file](https://s3.console.aws.amazon.com/s3/object/open-jobs-lake?region=eu-west-1&prefix=metaflow/metaflow-logs/1627390365/logs.txt) which corresponds to a metaflowtask running flow `good_overrides/example_flow.py`.

### CODE REVIEW

**Comment by jaklinger** [comment]
Merging since it's quite uncontroversial, and to speed up development in OJO it would help to have this in `dev`. I would however appreciate a review (at your convenience) - and we can roll back accordingly if there are issues.

File: daps_utils/docker_utils.py [hunk 1]
```diff
@@ -87,7 +87,8 @@ def format_logs(logs):
     """Join and format the logs"""
     logs = map(remove_ansi, logs)  # Remove ansi
     logs = map(lambda line: line.strip(b"\r"), logs)  # Remove trailing '\r'
-    logs = map(lambda line: b">>> %a" % line.decode(), logs)  # Prepend '>>> '
```
Comment (bishax) [inline daps_utils/docker_utils.py:90]: This set of maps would be more readable as a map over one function

```python
def _format_line(log):  # vanilla
    line = remove_ansi(log).strip(b"\r")
    return f">>> '{line.decode()}'".encode()

def _format_line(log):  # toolz
    return toolz.pipe(
        log,
        remove_ansi,
        bytes.strip(b"\r"),
        lambda line: f">>> '{line.decode()}'",
        str.encode,
    )

def format_logs(logs):
    return b"\n".join(format_line(log) for log in logs))
```

Comment (bishax) [reply daps_utils/docker_utils.py:90]: Whats the reason for decoding to prepend `>>> ` and then re-encoding?

**Review by bishax — COMMENTED** [review]
Looks good.

One minor nitpick and a question about why there's an extra step.

File: daps_utils/docker_utils.py (hunk 1 above)
Comment (jaklinger) [reply daps_utils/docker_utils.py:90]: Only to distinguish the contents of the `MetaflowTask` container logs from the task-level logs, but it's not particularly necessary. Do you reckon to drop it?

Comment (bishax) [reply daps_utils/docker_utils.py:90]: I meant why is it

decode `bytes -> str`; prepend ">>> "; encode `str -> bytes`

not just

prepend b">>> "

Comment (jaklinger) [reply daps_utils/docker_utils.py:90]: indeed, it could just be: `list(map(b">>> ".__add__, logs))`

Comment (jaklinger) [reply daps_utils/docker_utils.py:90]: and so:

```python
def _format_line(log):  # toolz
    return toolz.pipe(
        log,
        remove_ansi,
        lambda line: line.strip(b"\r"),
        b">>> ".__add__,
    )
```

Comment (bishax) [reply daps_utils/docker_utils.py:90]: That's a relief, I was worried I misunderstood something subtle but fundamental  :sweat_smile:

**Pull request merged**